src/zjs_sensor.c

```diff
@@ -0,0 +1,512 @@
+        return false;
+    }
+
+    // block until reply or timeout, we shouldn't see the ARC
+    // time out, if the ARC response comes back after it
+    // times out, it could pollute the result on the stack
+    if (!nano_sem_take(&sensor_sem, ZJS_SENSOR_TIMEOUT_TICKS)) {
+        PRINT("zjs_sensor_ipm_send_sync: FATAL ERROR, ipm timed out\n");
+        return false;
+    }
+
+    return true;
+}
+
+static int zjs_sensor_call_remote_function(zjs_ipm_message_t* send)
+{
+    zjs_ipm_message_t reply;
+    if (!zjs_sensor_ipm_send_sync(send, &reply)) {
+        return zjs_error("zjs_sensor_call_remote_function: ipm message failed or timed out!");
+    }
+    if (reply.error_code != ERROR_IPM_NONE) {
+        PRINT("zjs_sensor_call_remote_function: error code: %lu\n",
+              reply.error_code);
+    }
+    return reply.error_code;
+}
+
+static enum sensor_state zjs_sensor_get_state(jerry_value_t obj)
+{
+    const int BUFLEN = 20;
+    char buffer[BUFLEN];
+    if (zjs_obj_get_string(obj, "state", buffer, BUFLEN)) {
+        if (!strcmp(buffer, "idle"))
+            return SENSOR_STATE_IDLE;
+        if (!strcmp(buffer, "activating"))
+            return SENSOR_STATE_ACTIVATING;
+        if (!strcmp(buffer, "activated"))
+            return SENSOR_STATE_ACTIVATED;
+        if (!strcmp(buffer, "errored"))
+            return SENSOR_STATE_ERRORED;
+    }
+
+    PRINT("zjs_sensor_get_state: state not set\n");
+    jerry_value_t state = jerry_create_string("errored");
+    zjs_set_property(obj, "state", state);
+    jerry_release_value(state);
+    return SENSOR_STATE_ERRORED;
+}
+
+static void zjs_sensor_set_state(jerry_value_t obj, enum sensor_state state)
+{
+    enum sensor_state old_state = zjs_sensor_get_state(obj);
+    if (old_state == state) {
+        return;
+    }
+
+    // update state property and trigger onstatechange event
+    jerry_value_t new_state;
+    switch(state) {
+    case SENSOR_STATE_IDLE:
+        new_state = jerry_create_string("idle");
+        break;
+    case SENSOR_STATE_ACTIVATING:
+        new_state = jerry_create_string("activating");
+        break;
+    case SENSOR_STATE_ACTIVATED:
+        new_state = jerry_create_string("activated");
+        break;
+    case SENSOR_STATE_ERRORED:
+        new_state = jerry_create_string("errored");
+        break;
+    }
+    zjs_set_property(obj, "state", new_state);
+
+    jerry_value_t func = zjs_get_property(obj, "onstatechange");
+    if (jerry_value_is_function(func)) {
+        // if onstatechange exists, call it
+        jerry_value_t rval = jerry_call_function(func, obj, &new_state, 1);
+        if (jerry_value_has_error_flag(rval)) {
+            PRINT("zjs_sensor_set_state: error calling onstatechange\n");
+        }
+        jerry_release_value(rval);
+    }
+    jerry_release_value(func);
+
+    if (old_state == SENSOR_STATE_ACTIVATING &&
+        state == SENSOR_STATE_ACTIVATED) {
+        func = zjs_get_property(obj, "onactivate");
```

Comment:
Great that you are actively following the spec, this was just added two days ago :-)

https://github.com/w3c/sensors/commit/a85830bef0566b21176ac5af1ba0b1d8241b06ca